lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, Any]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
```

Comment:
Does this `__init__` method do anything? The line typing `_query_params` above should already set it for us

Reply:
Hmm without it, it goes into an infinite loop because:
1.  it calls `__setattr__` when we initialize `_query_params` 
2. we redefined `__setattr__` to access `_query_params` but it's not defined yet.

Here's a stackoverflow post that makes it seem like it's a common issue: https://stackoverflow.com/a/25179980

Reply:
Yeah, that's what the comment was previously added to explain. The link to stackoverflow doesn't add any additional clarity to the explanation that's already here.

The question I'm asking is whether we can just rely on `_query_params` getting automatically set by the `dataclasses` machinery (that is, whether it bypasses the issue with `__setattr__`). It looks like the answer is no, so we'd want to modify the comment to explain that we can't rely on writing `_query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)` to set this for us.

We should also just change the line above to `_query_params: Dict[str, Union[List[str], str]]` since the auto-field-setting doesn't actually do anything.

Reply:
This is most likely all entirely unimportant given that we can dodge these issues entirely by just not defining the attr-based methods in this class at all.